File: leetcode/895.最大频率栈.go
type FreqStack struct {
    cnt map[int]int
    stacks [][]int
}


func Constructor() FreqStack {
    return FreqStack{cnt: map[int]int{}}
}


func (f *FreqStack) Push(val int)  {
    c := f.cnt[val]
    if c == len(f.stacks){
        f.stacks = append(f.stacks, []int{val})
    }else{
        f.stacks[c] = append(f.stacks[c], val)
    }
    f.cnt[val]++
}


func (f *FreqStack) Pop() int {
    back := len(f.stacks) - 1
    st := f.stacks[back]
    bk := len(st) - 1
    val := st[bk]
    if bk == 0{
        f.stacks = f.stacks[:back]
    }else{
        f.stacks[back] = st[:bk]
    }
    f.cnt[val]--
    return val
}


/**
 * Your FreqStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * param_2 := obj.Pop();
 */

File: leetcode/895.最大频率栈.java
class FreqStack {
    private final Map<Integer, Integer> cnt = new HashMap<>();
    private final List<Deque<Integer>> stacks = new ArrayList<>();
    public void push(int val) {
        int c = cnt.getOrDefault(val, 0);
        if(c == stacks.size()){
            stacks.add(new ArrayDeque<>());
        }
        stacks.get(c).push(val);
        cnt.put(val, c+1);
    }
    
    public int pop() {
        int back = stacks.size() - 1;
        int val = stacks.get(back).pop();
        if(stacks.get(back).isEmpty()){
            stacks.remove(back);
        }
        cnt.put(val, cnt.get(val) - 1);
        return val;
    }
}

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(val);
 * int param_2 = obj.pop();
 */